### PR TITLE
Parse PEP 484 type comments

### LINKED ISSUE #247: Detect type hint comments: PEP484
*Original report by [w-thomas-q](https://launchpad.net/~w-thomas-q) on [Launchpad](https://bugs.launchpad.net/bugs/1587170):*

------------------------------------

Hi,

thanks for this really great tool!
One issues I stumbled upon was that type hints seem to be ignored / not detected.
For example, we have an import at the top of a file:

    from typing import List

And then later in an `__init__` function something like this:

    self.scripts = []  # type: List[script.Script]

Now running pyflakes (through flake8) gives me this error:

    foo.py:4:1: F401 'List' imported but unused

PEP484 defines this kind of type comments.
Is there an option to enable type hint detection?
Or can we have that as feature request? I guess I'm fine with type comments only for the time being, but of course a fully PEP484 compliant solution would be greatly appreciated :)

Thanks!

### PR DESCRIPTION
**Pull request by asottile**
Resolves #247

### CODE REVIEW

**Comment by asottile** [comment]
some of the patches here are a little unrelated, let me know and I can split them out into separate PRs to focus on the changes here:
- `setup.py test` -> `-m unittest discovery pyflakes`
    - I had some major problems getting flake8 to install with `pyflakes.egg-info` sitting around from the `setup.py test` invocations.  after resolving those, I _think_ this was unnecessary but imo an improvement
- `compile(4 arg)` -> `ast.parse(...)` -- completely unrelated, but simplified some things
- modern flake8 - I'm using a not-yet-released version which contains [this patch](https://gitlab.com/pycqa/flake8/commit/a1532441917ab9ea70838903095ffe1d802e541c)

File: pyflakes/checker.py [hunk 1]
```diff
@@ -518,6 +527,60 @@ def is_typing_overload_decorator(node):
...
+
+
+class _TypeableVisitor(ast.NodeVisitor):
+    """Collect the line number and nodes which are deemed typeable by
+    PEP 484
+
+    https://www.python.org/dev/peps/pep-0484/#type-comments
+    """
+    def __init__(self):
+        self.typeable_lines = []  # type: int
```
Comment (sigmavirus24) [inline pyflakes/checker.py]: Should tihs be `# type: List[int]`?

File: pyflakes/checker.py [hunk 2]
```diff
@@ -518,6 +527,60 @@ def is_typing_overload_decorator(node):
...
+        self.typeable_nodes = {}  # type: Dict[int, ast.AST]
+
+    def _typeable(self, node):
+        # if there is more than one typeable thing on a line last one wins
+        self.typeable_lines.append(node.lineno)
```
Comment (sigmavirus24) [inline pyflakes/checker.py]: Perhaps this is silly, but why are we doubly tracking the line number? Could `typeable_lines` simply be a property that does `list(self.typeable_nodes.values())`?

File: pyflakes/checker.py [hunk 3]
```diff
@@ -518,6 +527,60 @@ def is_typing_overload_decorator(node):
...
+def _collect_type_comments(tree, tokens):
+    visitor = _TypeableVisitor()
+    visitor.visit(tree)
+
+    type_comments = collections.defaultdict(list)
+    for tp, text, start, _, _ in tokens:
+        if (
+                tp != tokenize.COMMENT or  # skip non comments
+                not text.startswith('# type:') or  # skip non-type comments
+                text.split()[:3] == ['#', 'type:', 'ignore']  # skip ignores
```
Comment (sigmavirus24) [inline pyflakes/checker.py]: Perhaps another silly question, why are we doing `text.split()[:3]` to check this but not using split above to see if the comment starts with `# type:`?

My guess is that the split here is to accommodate multiple spaces so that all of the following work:

```
# type: ignore
#   type:       ignore
# type:    ignore
#      type: ignore
```

But then we're going to bail on those lines anyway because of `not text.startswith('# type:')` so I'd like to understand the decision here. Do we perhaps want to do:

```py
for tp, text, start, _, _ in tokens:
    text_pieces = text.split()
    if (
        tp != tokenize.COMMENT or
        not (text_pieces[:2] == ['#', 'type:'] and
             text_pieces[3] == 'ignore')
    ):
        comment
```

Comment (sigmavirus24) [reply pyflakes/checker.py]: I don't know the intricacies of mypy but does it support `#type:ignore` as well? Do we need to handle that? Should this be a regular expression?

File: pyflakes/checker.py [hunk 4]
```diff
@@ -948,7 +1013,25 @@ def on_conditional_branch():
             except KeyError:
                 self.report(messages.UndefinedName, node, name)
 
+    def _handle_type_comments(self, node):
+        for (lineno, col_offset), comment in self._type_comments.get(node, ()):
+            comment = comment[len('# type:'):].strip()
+            func_match = re.match(r'^(\(.*?\)) *-> *(.*)$', comment)
```
Comment (sigmavirus24) [inline pyflakes/checker.py]: If I remember correctly, compiling the regular expression can be expensive. Should we do that once at the module level and re-use it over-and-over instead of compiling each time?

File: pyflakes/checker.py (hunk 1 above)
Comment (asottile) [reply pyflakes/checker.py]: yes, yes it should -- I should write less code while celebrating new years :laughing:

File: pyflakes/checker.py (hunk 2 above)
Comment (asottile) [reply pyflakes/checker.py]: I couldn't think of a way to make the lookup efficient without needing two state variables here.  having a property which sorts the keys would invoke a O(N) (`list(typeable_nodes)`) + O(N*log(N)) (`sorted(...)`) + O(log(N)) (`bisect.bisect_right(...)`) instead of just the `O(log(N))` lookup

File: pyflakes/checker.py (hunk 3 above)
Comment (asottile) [reply pyflakes/checker.py]: yeah looks like there's more to support here, I'll make a regex for this.  Even this is valid: `#type:()->None`

File: pyflakes/checker.py (hunk 4 above)
Comment (asottile) [reply pyflakes/checker.py]: you are correct, I thought I had hoisted these but looks like I forgot one!

File: pyflakes/checker.py (hunk 2 above)
Comment (sigmavirus24) [reply pyflakes/checker.py]: Makes sense. :+1:

File: pyflakes/checker.py [hunk 5]
```diff
@@ -552,7 +622,8 @@ class Checker(object):
         builtIns.update(_customBuiltIns.split(','))
     del _customBuiltIns
 
-    def __init__(self, tree, filename='(none)', builtins=None,
+    # TODO: tokens=() is just to make flake8 happy
+    def __init__(self, tree, tokens=(), filename='(none)', builtins=None,
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: if I move `tokens=()` the api is consistent, though what I really would want is `tokens` to be a required argument.  that said, if it is left off pyflakes will still lint -- just it won't check type comments

File: pyflakes/test/test_checker.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        self.assertSetEqual(ret, set())
+
+    def test_type_ignore_comment_ignored(self):
+        """
+        Test that `# type: ignore` comments are not collected.
+        """
+        ret = self._collect('x = 1  # type: ignore')
+        self.assertSetEqual(ret, set())
+
+    def test_type_ignore_with_other_things_ignored(self):
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/test/test_checker.py]: will do, it's not exactly an interesting case from the implementation's standpoint though -- the only reason this is here was to regress-test the edge-case of `ignore` parsing -- I can add this nonetheless

File: pyflakes/test/test_other.py [hunk 7]
```diff
@@ -2073,6 +2073,45 @@ class A:
...
+
+        x = 1  #type:F
+        """)
+
+    def test_typeCommentsInvalidDoesNotMarkAsUsed(self):
+        self.flakes("""
+        from mod import F
+
+        # type: F
+        """, m.UnusedImport)
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/test/test_other.py]: yep this is intentional

File: pyflakes/checker.py [hunk 8]
```diff
@@ -518,6 +534,60 @@ def is_typing_overload_decorator(node):
...
+    for tp, text, start, _, _ in tokens:
+        if (
+                tp != tokenize.COMMENT or  # skip non comments
+                not TYPE_COMMENT_RE.match(text) or  # skip non-type comments
+                TYPE_IGNORE_RE.match(text)  # skip ignores
+        ):
+            continue
+
+        lineno, _ = start
+        idx = bisect.bisect_right(visitor.typeable_lines, lineno)
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: my goal with "associate to previous" was to get the proper scope in which it has an effect such that pyflakes can check the right variables there.  you're right this example doesn't quite make sense and that `typed-ast` errors on this.  I couldn't quite make this happen with just cross referencing.

The actual reason it's "look backwards" is for functions which are annotated as follows:

```python
def f(
    a,  # type: int  # (annotation for a specific argument)
):
    # type: (...) -> None # (annotation for the function's return value)
    pass
```

In this case it's necessary to assign both the arguments and the type signature to the function itself, and make sure to evaluate those in the module scope instead of the function scope

File: pyflakes/test/test_checker.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+    def test_last_node_wins(self):
+        """
+        Test that when two typeable nodes are present on a line, the last
+        typeable one wins.
+        """
+        ret = self._collect('def f(): x = 1  # type: int')
+        self.assertSetEqual(ret, {(ast.Assign, ('# type: int',))})
+
+    def test_function_def_assigned_comments(self):
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/test/test_checker.py]: It should work, not by accident (I can add a test for this)

`(str, List[int])` is a tuple expression, which itself is an expression statement

File: pyflakes/checker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: this'll require two branches instead of 1 -- kind of a leaky abstraction.  I'll add a comment instead if it's not clear

File: pyflakes/test/test_other.py [hunk 10]
```diff
@@ -2073,6 +2073,45 @@ class A:
         class B: pass
         ''', m.UndefinedName)
 
+    def test_typeCommentsMarkImportsAsUsed(self):
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/test/test_other.py]: hah oh boy this class got abused for all sorts of things -- I just stuck this next to the other type annotation tests.

How about I split this up in a followup? there's a lot of diff churn already here and I think it would make sense for just a "mv" commit that's easy to review

File: pyflakes/checker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: pyflakes has an `api.py` file which I suspect _should_ be the public api, however even flake8 reaches into pyflakes/checker.py :man_facepalming:

File: pyflakes/test/test_other.py [hunk 11]
```diff
@@ -2073,6 +2073,45 @@ class A:
...
+        from mod import F
+
+        # type: F
+        """, m.UnusedImport)
+
+    def test_typeCommentsSyntaxError(self):
+        self.flakes("""
+        def f(x):  # type: (F[) -> None
+            pass
+        """, m.ForwardAnnotationSyntaxError)
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/test/test_other.py]: urgh but then I need to thread that through -- I'll do this but it breaks the abstraction

File: pyflakes/checker.py [hunk 12]
```diff
@@ -552,7 +622,8 @@ class Checker(object):
         builtIns.update(_customBuiltIns.split(','))
     del _customBuiltIns
 
-    def __init__(self, tree, filename='(none)', builtins=None,
+    # TODO: tokens=() is just to make flake8 happy
```
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: yeah this is api shape issue -- I'm a bit stuck here on the right thing to do. Here's what I want and the constraints:

- tree + tokens are required position arguments -- they are necessary to perform all of the checks as implemented in this PR
- flake8 <=3.6.0 calls `Checker` directly with positional arguments (in 3.7.0+ it will call it with tree positionally and `filename=` as a named argument)
- pyflakes uses flake8 to validate itself

I *think* the easiest way forward is to soft-break the api for those which are calling `Checker()` without named arguments and release a minor version with `tokens=()`.  flake8 pins a specific version of pyflakes and can therefore update in lockstep safely.  At a later date, pyfalkes makes `tokens` a required positional argument and releases a major release.

Alternatively, we could just hard-break today and I could change the tests to create separate flake8 environments which don't use the source tree's pyflakes to lint things -- though this is much messier than what I have right now.

I think for option 1, the branch as-is is fine, though I should update the comment

File: pyflakes/checker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: going to resolve this discussion -- let's continue it in the one where I posted the big blob about flake8

File: pyflakes/test/test_other.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/test/test_other.py]: Created followup issue here: https://github.com/PyCQA/pyflakes/issues/402

File: pyflakes/checker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: I'm adding a test and comment comment demonstrating and explaining this limitation.

File: pyflakes/checker.py [hunk 13]
```diff
@@ -518,6 +534,63 @@ def is_typing_overload_decorator(node):
...
+
+def _collect_type_comments(tree, tokens):
+    visitor = _TypeableVisitor()
+    visitor.visit(tree)
+
+    type_comments = collections.defaultdict(list)
+    for tp, text, start, _, _ in tokens:
+        if (
+                tp != tokenize.COMMENT or  # skip non comments
+                not TYPE_COMMENT_RE.match(text) or  # skip non-type comments
```
Comment (sigmavirus24) [inline pyflakes/checker.py]: Do we want match or search? I wonder, can mypy/typed_ast handle something like:


```py
foo = 1   # noqa: E11  # type: ignore
```

Comment (asottile) [reply pyflakes/checker.py]: typed-ast is picky and requires the type comment to come first:

(using [astpretty](https://github.com/asottile/astpretty))

```
$ cat t.py 
x = 1  # noqa  # type: int
$ astpretty --typed-27 t.py 
Module(
    body=[
        Assign(
            lineno=1,
            col_offset=0,
            targets=[Name(lineno=1, col_offset=0, id='x', ctx=Store())],
            value=Num(lineno=1, col_offset=4, n=1),
            type_comment=None,
        ),
    ],
    type_ignores=[],
)
```

```
$ cat t2.py 
x = 1  # type: int  # noqa
$ astpretty --typed-27 t2.py
Module(
    body=[
        Assign(
            lineno=1,
            col_offset=0,
            targets=[Name(lineno=1, col_offset=0, id='x', ctx=Store())],
            value=Num(lineno=1, col_offset=4, n=1),
            type_comment='int  # noqa',
        ),
    ],
    type_ignores=[],
)
```

Comment (sigmavirus24) [reply pyflakes/checker.py]: Sorry, you should look a bit more closely at the second AST you pretty-printed, namely:

```
            type_comment='int  # noqa',
```

In other words, typed AST doesn't handle any case where some other linter's comment-based directives are present.

Comment (asottile) [reply pyflakes/checker.py]: it works, the inner comments get discarded since they're re-parsed by the ast:

```
$ cat t.py
from typing import List
x = []  # type: List[int]  # noqa
reveal_type(x)
$ mypy t.py
t.py:3: error: Revealed type is 'builtins.list[builtins.int]'
```

**Comment by asottile** [comment]
I rebased to resolve the conflict with #398

File: pyflakes/checker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (asottile) [reply pyflakes/checker.py]: yeah I thought about warnings, but there's no precedent in pyflakes currently.  basically I'd do

```python
if not tokens:
    warnings.warn(FutureWarning, 'tokens was not specified, not checking `# type:` comments')
```

though this might be annoying for those that can't control their environments

**Comment by asottile** [comment]
@PeterJCLaw did you get a chance to try this on larger codebases?  I tried it against a few of our larger codebases at lyft with success

**Comment by PeterJCLaw** [comment]
@asottile yup, just done that. Looks good (even found a few places where we had redundant imports :)

File: pyflakes/checker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (PeterJCLaw) [reply pyflakes/checker.py]: Ah, good point about this ending up showing for end users rather more than the developers of the intermediate packages.
I'm not sure what's best in that case. Perhaps just something in the docs?

File: tox.ini [hunk 14]
```diff
@@ -6,11 +6,10 @@ envlist =
 [testenv]
 deps = flake8==3.6.0
 commands =
-    python setup.py test -q
+    python -m unittest discover pyflakes
+    flake8 --version
```
Comment (sigmavirus24) [inline tox.ini]: Is this left over from debugging?

Comment (asottile) [reply tox.ini]: seems it was an implicit conflict with https://github.com/PyCQA/pyflakes/pull/398/files#diff-b91f3d5bd63fcd17221b267e851608e8L10 from #398

**Comment by asottile** [comment]
rebased and removed the `flake8 --version`

**Comment by sigmavirus24** [comment]
This is awesome! Thanks @asottile!

**Comment by asottile** [comment]
no problem! pretty excited about this change myself, it's really all that's stopping me from using mypy in 2+3 projects

**Comment by sigmavirus24** [comment]
I'd love to start taking advantage of this in Flake8 proper.

**Comment by asottile** [comment]
If pyflakes makes a release, I think all that's blocking flake8 3.7.0 would be the fixing of `per-file-ignores` 🎉

**Comment by jriddy** [comment]
Do you guys know when you plan to cut a release with this in it?
